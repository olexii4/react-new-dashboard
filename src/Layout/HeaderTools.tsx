/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import gravatarUrl from 'gravatar-url';
import {
  Avatar,
  Dropdown,
  DropdownItem,
  DropdownToggle,
  PageHeaderTools,
  PageHeaderToolsGroup,
  PageHeaderToolsItem,
} from '@patternfly/react-core';

import { ThemeVariant } from './themeVariant';

type Props = {
  userEmail: string;
  userName: string;
  logout: () => void;
  changeTheme: (theme: ThemeVariant) => void;
};
type State = {
  isOpen: boolean;
}

export class HeaderTools extends React.PureComponent<Props, State> {

  constructor(props: Props) {
    super(props);

    this.state = {
      isOpen: false,
    };
  }

  private onSelect(): void {
    this.setState({
      isOpen: !this.state.isOpen,
    });
  }

  private onToggle(isOpen: boolean): void {
    this.setState({ isOpen });
  }

  private setTheme(theme: ThemeVariant): void {
    this.props.changeTheme(theme);
  }

  private buildDropdownItems(): Array<React.ReactElement> {
    return [
      (
        <DropdownItem
          key='light'
          component='button'
          onClick={(): void => this.setTheme(ThemeVariant.LIGHT)}
        >
          Light Theme
        </DropdownItem>
      ),
      (
        <DropdownItem
          key='dark'
          component='button'
          onClick={(): void => this.setTheme(ThemeVariant.DARK)}
        >
          Dark Theme
        </DropdownItem>
      ),
      (
        <DropdownItem
          key='account_logout'
          component='button'
          onClick={() => this.props.logout()}
        >
          Logout
        </DropdownItem>
      )
    ];
  }

  private buildToggleButton(): React.ReactElement {
    return (
      <DropdownToggle onToggle={isOpen => this.onToggle(isOpen)}>
        {this.props.userName}
      </DropdownToggle>
    );
  }

  public render(): React.ReactElement {
    const { isOpen } = this.state;

    const imageUrl = gravatarUrl(this.props.userEmail, { default: 'retro' });
    const avatar = <Avatar src={imageUrl} alt='Avatar image' />;

    const toggleButton = this.buildToggleButton();
    const dropdownItems = this.buildDropdownItems();

    return (
      <PageHeaderTools>
        <PageHeaderToolsGroup>
          <PageHeaderToolsItem>
            <Dropdown
              isPlain
              position="right"
              onSelect={() => this.onSelect()}
              isOpen={isOpen}
              toggle={toggleButton}
              dropdownItems={dropdownItems}
            />
          </PageHeaderToolsItem>
        </PageHeaderToolsGroup>
        {avatar}
      </PageHeaderTools>
    );
  }

}
