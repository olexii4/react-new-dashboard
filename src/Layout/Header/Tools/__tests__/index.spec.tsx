/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import HeaderTools from '..';

jest.mock('gravatar-url', () => {
  return function () {
    return 'avatar/source/location';
  };
});

describe('Page header tools', () => {
  const mockLogout = jest.fn();
  const mockChangeTheme = jest.fn();

  const email = 'johndoe@example.com';
  const name = 'John Doe';

  const component = (<HeaderTools
    userEmail={email}
    userName={name}
    logout={mockLogout}
    changeTheme={mockChangeTheme}
  />);

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should correctly render the component', () => {
    expect(renderer.create(component).toJSON()).toMatchSnapshot();
  });

  it('should open the dropdown', () => {
    render(component);

    const menuButton = screen.getByRole('button', { name });
    fireEvent.click(menuButton);

    const items = screen.getAllByRole('menuitem');
    expect(items.length).toEqual(3);
  });

  it('should fire the logout event', () => {
    render(component);

    const menuButton = screen.getByRole('button', { name });
    fireEvent.click(menuButton);

    const logoutItem = screen.getByRole('menuitem', { name: /logout/i });
    fireEvent.click(logoutItem);

    expect(mockLogout).toBeCalled();
  });

});
