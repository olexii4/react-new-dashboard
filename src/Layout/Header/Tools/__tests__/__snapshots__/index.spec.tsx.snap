// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page header tools should correctly render the component 1`] = `
<div
  className="pf-c-page__header-tools"
>
  <div
    className="pf-c-page__header-tools-group"
  >
    <div
      className="pf-c-page__header-tools-item"
    >
      <div
        className="pf-c-dropdown pf-m-align-right"
        data-ouia-component-id={0}
        data-ouia-component-type="PF4/Dropdown"
        data-ouia-safe={true}
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="pf-c-dropdown__toggle pf-m-plain"
          disabled={false}
          id="pf-toggle-id-0"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span
            className="pf-c-dropdown__toggle-text"
          >
            John Doe
          </span>
          <span
            className="pf-c-dropdown__toggle-icon"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                transform=""
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
  <img
    alt="Avatar image"
    className="pf-c-avatar"
    src="avatar/source/location"
  />
</div>
`;
